spec/unit/provider/group/gpasswd_spec.rb

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env rspec
+require 'spec_helper'
+
+describe Puppet::Type.type(:group).provider(:gpasswd), :unless => Puppet.features.microsoft_windows? do
+  before do
+    described_class.stubs(:command).with(:add).returns '/usr/sbin/groupadd'
+    described_class.stubs(:command).with(:delete).returns '/usr/sbin/groupdel'
+    described_class.stubs(:command).with(:modify).returns '/usr/sbin/groupmod'
+    described_class.stubs(:command).with(:addmember).returns '/usr/bin/gpasswd'
+    described_class.stubs(:command).with(:delmember).returns '/usr/bin/gpasswd'
+  end
+
+  let(:resource) { Puppet::Type.type(:group).new(:name => 'mygroup', :provider => provider) }
+  let(:provider) { described_class.new(:name => 'mygroup') }
+
+  describe "#create" do
+    it "should add -o when allowdupe is enabled and the group is being created" do
+      resource[:allowdupe] = :true
+      provider.expects(:execute).with(
+        '/usr/sbin/groupadd -o mygroup',
+        # This is hard coded into the 'nameservice' provider so this isn't a
+        # great test, but I'm not sure how to override it properly.
+        {:failonfail => true, :combine => true, :custom_environment => {}}
+      )
+      provider.create
+    end
+
+    describe "on system that feature system_groups", :if => described_class.system_groups? do
+      it "should add -r when system is enabled and the group is being created" do
+        resource[:system] = :true
+        provider.expects(:execute).with(
+          '/usr/sbin/groupadd -r mygroup',
+          # This is hard coded into the 'nameservice' provider so this isn't a
+          # great test, but I'm not sure how to override it properly.
+          {:failonfail => true, :combine => true, :custom_environment => {}}
+        )
+        provider.create
+      end
+    end
+
+    describe "on system that do not feature system_groups", :unless => described_class.system_groups? do
+      it "should not add -r when system is enabled and the group is being created" do
+        resource[:system] = :true
+        provider.expects(:execute).with(
+          '/usr/sbin/groupadd mygroup',
+          # This is hard coded into the 'nameservice' provider so this isn't a
+          # great test, but I'm not sure how to override it properly.
+          {:failonfail => true, :combine => true, :custom_environment => {}}
+        )
+        provider.create
+      end
+    end
+
+    describe "when adding additional group members to the group" do
+      it "should pass all members individually as group add options to gpasswd" do
+        resource[:members] = ['test_one','test_two','test_three']
+        provider.expects(:execute).with() { |value|
+          value.split(' && ').map{ |x|
+            if x =~ /.*\/usr\/bin\/gpasswd -a (.*) mygroup.*/
+              x = $1
+            end
+          }.compact.sort.should == resource[:members].sort
```

Comment:
We recently updated to rspec3, so rspec-expectations no longer defines the `#should` method on `Kernel`. See http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/ for more info. So something like the following should work

```
provider.expects(:execute).with do |value|
  members = value.split(' && ').map do |x|
              ...
            end.compact.sort
  expect(members).to eq(resource[:members].sort)
end
```